sql/src/main/java/io/crate/operation/projectors/ShardDMLExecutor.java

```diff
@@ -0,0 +1,200 @@
+    implements Function<BatchIterator, CompletableFuture<? extends Iterable<Row>>> {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardDMLExecutor.class);
+
+    private static final BackoffPolicy BACKOFF_POLICY = LimitedExponentialBackoff.limitedExponential(1000);
+    public static final int DEFAULT_BULK_SIZE = 10_000;
+
+    private final int bulkSize;
+    private final ScheduledExecutorService scheduler;
+    private final BitSet responses;
+    private final ClusterService clusterService;
+    private final NodeJobsCounter nodeJobsCounter;
+    private final AtomicInteger pendingItemsCount = new AtomicInteger(0);
+    private final Consumer<Row> rowConsumer;
+    private final BooleanSupplier backpressureTrigger;
+    private final Function<Boolean, CompletableFuture<BitSet>> execute;
+
+    private TReq currentRequest;
+    private int numItems = -1;
+    private CompletableFuture executionFuture = new CompletableFuture();
+
+    public ShardDMLExecutor(int bulkSize,
+                            ScheduledExecutorService scheduler,
+                            CollectExpression<Row, ?> uidExpression,
+                            ClusterService clusterService,
+                            NodeJobsCounter nodeJobsCounter,
+                            Supplier<TReq> requestFactory,
+                            Function<String, TItem> itemFactory,
+                            BiConsumer<TReq, ActionListener<ShardResponse>> transportAction) {
+        this.bulkSize = bulkSize;
+        this.scheduler = scheduler;
+        this.clusterService = clusterService;
+        this.nodeJobsCounter = nodeJobsCounter;
+        this.responses = new BitSet();
+        this.currentRequest = requestFactory.get();
+
+        this.rowConsumer = createRowConsumer(uidExpression, itemFactory);
+        this.backpressureTrigger = createBackpressureTrigger();
+        this.execute = createExecuteFunction(scheduler, nodeJobsCounter, requestFactory, transportAction);
+    }
+
+    private Consumer<Row> createRowConsumer(CollectExpression<Row, ?> uidExpression,
+                                            Function<String, TItem> itemFactory) {
+        return (row) -> {
+            numItems++;
+            uidExpression.setNextRow(row);
+            currentRequest.add(numItems, itemFactory.apply(((BytesRef) uidExpression.value()).utf8ToString()));
+        };
+    }
+
+    private BooleanSupplier createBackpressureTrigger() {
+        return () -> {
+            final String nodeId = getLocalNodeId();
+            if (isExecutionPossible(nodeId)) {
+                return false;
+            }
+            return true;
+        };
+    }
+
+    private boolean isExecutionPossible(String nodeId) {
+        return nodeJobsCounter.getInProgressJobsForNode(nodeId) < MAX_NODE_CONCURRENT_OPERATIONS;
+    }
+
+    private Function<Boolean, CompletableFuture<BitSet>> createExecuteFunction(ScheduledExecutorService scheduler,
```

Comment:
I would appreciate some comments on this method. (inside the body, I'm quite confused :-( )